docs/madr/decisions/011-mesh-traffic-route.md

```diff
@@ -0,0 +1,964 @@
+
+The new route matches requests based on:
+
+- headers
+- path
+- method
+- query parameters
+
+##### Headers
+
+Headers can be tested for:
+
+- presence
+- absence
+- exact value
+- prefix value
+- regex value
+
+##### Spec
+
+A rule contains a list of match objects, each of which specifies a match for one
+or more of the above types:
+
+```yaml
+matches:
+ - method: POST
+   path:
+    type: RegularExpression|Exact|Prefix
+    value: ...
+   headers:
+    - type: RegularExpression|Exact|Prefix|Absent
+      name: "..."
+      value: ...
+   queryParams:
+    - type: RegularExpression|Exact
+      name: "..."
+      value: ...
+```
+
+A given entry in `match` succeeds if _all_ of the conditions are satisfied and the rule matches when _any one_ of the entries under `match` succeeds. In other words, `match` is an OR'd list of AND'd conditions.
+
+#### Request modification
+
+With `filters` we can change:
+
+- request headers
+- response headers
+- request mirroring (response is ignored)
+- HTTP redirect
+- host
+- request path
+- the response by directly returning one
+
+```yaml
+filters:
+ - type: RequestHeaderModifier|ResponseHeaderModifier|RequestMirror|RequestRedirect|URLRewrite|DirectResponse
+   requestHeaderModifier:
+    set:
+     - name:
+       value:
+    add:
+     - name:
+       value:
+    remove:
+     - "name"
+   responseHeaderModifier:
+    set:
+     - name:
+       value:
+    add:
+     - name:
+       value:
+    remove:
+     - name
+   requestMirror:
+    backendRef:
+     kind: MeshService
+     name: backend-mirror
+   requestRedirect:
+    scheme:
+    hostname:
+    path:
+    port:
+    statusCode:
+   urlRewrite:
+    hostname:
+    path:
+     replaceFullPath:
+     replacePrefixMatch:
+   directResponse:
+    status:
+    body:
+     inlineBytes: <base64>
+     inlineString:
+     filename:
+     environmentVariable:
+```
+
+#### Routing
+
+If `matches` succeeds, the request is routed to the specified weighted destinations.
+
+The destinations appear as references in `backendRefs`.
+In the case that a request should be routed to the service it would otherwise
+have been routed to, `backendRefs` can be omitted.
+
+```yaml
+- backendRefs:
+  - weight: 90
+    kind: MeshServiceSubset
+    name: backend
+    tags:
+      version: v2
+  - weight: 10
+    kind: MeshServiceSubset
+    name: backend
+    tags:
+      version: v2
+  ...
+```
+
+Note that it's not a valid configuration to have a `requestRedirect` and
+`backendRefs` specified or both `requestRedirect` and `urlRewrite`.
+
+Note that in Gateway API, filters are also available on a _per-backend_ basis (`HTTPBackendRef.filters`).
+We leave this option open for our policy in the future.
+
+#### Load balancing
+
+At the moment `TrafficRoute` can be used to configure load balancing behavior at
+the top, rule-independent level. This MADR proposes **removing** load balancing from
+the route policies and instead later creating a separate policy to handle load
```

Comment:
How are we going to reference an "envoy cluster" in this new load balancing policy?

Reply:
There's a 1:1 mapping between Envoy clusters and `kuma.io/service`

Reply:
Today there is 1:1 mapping between Route and Envoy clusters, we can generate many clusters for a single service:
```
backend_-0- for backend v1
backend_-1- for backend v2
```
But we discussed with @michaelbeaumont that you probably want to set a load balancer per `kuma.io/service` not per Envoy cluster. In that case it's fine.